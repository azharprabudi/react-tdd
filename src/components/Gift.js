import React, { PureComponent } from "react";
import {
  Form,
  FormGroup,
  FormControl,
  ControlLabel,
  Button
} from "react-bootstrap";

class Gift extends PureComponent {
  state = {
    person: "",
    present: ""
  };

  onChangeText = stateName => e => {
    this.setState({
      [stateName]: e.target.value
    });
  };

  onRemoveGift = () => {
    this.props.onRemoveGift(this.props.gift.id);
  };

  render() {
    return (
      <div>
        <Form>
          <FormGroup>
            <ControlLabel>Person</ControlLabel>
            <FormControl
              className="input-person"
              onChange={this.onChangeText("person")}
              value={this.state.person}
            />
          </FormGroup>
          <FormGroup>
            <ControlLabel>Present</ControlLabel>
            <FormControl
              className="input-present"
              onChange={this.onChangeText("present")}
              value={this.state.present}
            />
          </FormGroup>
          <FormGroup>
            <Button className="btn-remove-gift" onClick={this.onRemoveGift}>
              Delete
            </Button>
          </FormGroup>
        </Form>
      </div>
    );
  }
}

export default Gift;
