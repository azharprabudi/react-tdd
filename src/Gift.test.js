import React from "react";
import { shallow } from "enzyme";
import Gift from "./components/Gift";

describe("Gift", () => {
  const mockRemove = jest.fn();
  const idRemove = 1;
  const props = {
    gift: { person: "", present: "", id: idRemove },
    key: idRemove,
    onRemoveGift: mockRemove
  };

  const gift = shallow(<Gift {...props} />);

  it("it renders correctly", () => {
    expect(gift).toMatchSnapshot();
  });

  it("initialize the `state` with the empty person and present", () => {
    expect(gift.state()).toEqual({
      person: "",
      present: ""
    });
  });

  /* bdd when user change the input */

  describe("when user change the text input", () => {
    describe("change the person", () => {
      const personName = "azhar";
      beforeEach(() => {
        gift
          .find(".input-person")
          .simulate("change", { target: { value: personName } });
      });

      it(`the state should be ${personName}`, () => {
        expect(gift.state().person).toEqual(personName);
      });
    });

    describe("change the present", () => {
      const presentName = "tamiya";
      beforeEach(() => {
        gift
          .find(".input-present")
          .simulate("change", { target: { value: presentName } });
      });

      it(`the state should be ${presentName}`, () => {
        expect(gift.state().present).toEqual(presentName);
      });
    });
  });

  describe("when clicking `remove gift` button", () => {
    beforeEach(() => {
      gift.find(".btn-remove-gift").simulate("click");
    });

    it("call the onRemoveGift callback", () => {
      expect(mockRemove).toHaveBeenCalledWith(idRemove);
    });
  });
});
