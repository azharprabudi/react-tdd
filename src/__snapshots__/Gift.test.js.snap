// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift it renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Gift
    gift={
      Object {
        "id": 1,
        "person": "",
        "present": "",
      }
    }
    onRemoveGift={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Form
        bsClass="form"
        componentClass="form"
        horizontal={false}
        inline={false}
      >
        <FormGroup
          bsClass="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Person
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="input-person"
            componentClass="input"
            onChange={[Function]}
            value=""
          />
        </FormGroup>
        <FormGroup
          bsClass="form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Present
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="input-present"
            componentClass="input"
            onChange={[Function]}
            value=""
          />
        </FormGroup>
        <FormGroup
          bsClass="form-group"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-remove-gift"
            disabled={false}
            onClick={[Function]}
          >
            Delete
          </Button>
        </FormGroup>
      </Form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "form",
        "children": Array [
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Person
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-person"
              componentClass="input"
              onChange={[Function]}
              value=""
            />
          </FormGroup>,
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Present
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-present"
              componentClass="input"
              onChange={[Function]}
              value=""
            />
          </FormGroup>,
          <FormGroup
            bsClass="form-group"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-remove-gift"
              disabled={false}
              onClick={[Function]}
            >
              Delete
            </Button>
          </FormGroup>,
        ],
        "componentClass": "form",
        "horizontal": false,
        "inline": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Person
              </ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-person"
                componentClass="input"
                onChange={[Function]}
                value=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Person",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Person",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-person",
                "componentClass": "input",
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Present
              </ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-present"
                componentClass="input"
                onChange={[Function]}
                value=""
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Present",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Present",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-present",
                "componentClass": "input",
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-remove-gift"
              disabled={false}
              onClick={[Function]}
            >
              Delete
            </Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": "Delete",
              "className": "btn-remove-gift",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Delete",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Form
          bsClass="form"
          componentClass="form"
          horizontal={false}
          inline={false}
        >
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Person
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-person"
              componentClass="input"
              onChange={[Function]}
              value=""
            />
          </FormGroup>
          <FormGroup
            bsClass="form-group"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Present
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              className="input-present"
              componentClass="input"
              onChange={[Function]}
              value=""
            />
          </FormGroup>
          <FormGroup
            bsClass="form-group"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-remove-gift"
              disabled={false}
              onClick={[Function]}
            >
              Delete
            </Button>
          </FormGroup>
        </Form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form",
          "children": Array [
            <FormGroup
              bsClass="form-group"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Person
              </ControlLabel>
              <FormControl
                bsClass="form-control"
                className="input-person"
                componentClass="input"
                onChange={[Function]}
                value=""
              />
            </FormGroup>,
            <FormGroup
              bsClass="form-group"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Present
              </ControlLabel>
              <FormControl
                bsClass="form-control"
                className="input-present"
                componentClass="input"
                onChange={[Function]}
                value=""
              />
            </FormGroup>,
            <FormGroup
              bsClass="form-group"
            >
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                className="btn-remove-gift"
                disabled={false}
                onClick={[Function]}
              >
                Delete
              </Button>
            </FormGroup>,
          ],
          "componentClass": "form",
          "horizontal": false,
          "inline": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Person
                </ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-person"
                  componentClass="input"
                  onChange={[Function]}
                  value=""
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Person",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Person",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-person",
                  "componentClass": "input",
                  "onChange": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Present
                </ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-present"
                  componentClass="input"
                  onChange={[Function]}
                  value=""
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Present",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Present",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-present",
                  "componentClass": "input",
                  "onChange": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                className="btn-remove-gift"
                disabled={false}
                onClick={[Function]}
              >
                Delete
              </Button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "default",
                "children": "Delete",
                "className": "btn-remove-gift",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Delete",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
